specification/sbi_atee.adoc

```diff
@@ -10,8 +39,1296 @@ shown in the figure above, the TEE Runtime (operating in VS-mode) supports
+and the length is specified by by `region_len`. `tvm_memory_region_type` must be a
+legal value for the `tvm_memory_region_type` enum type described below.
+
+Both `tvm_gpa_addr` and `region_len` must be 4kB-aligned, and the region must not
+overlap with a previously defined region. This call must not be made after calling
+`sbi_tee_host_finalize_tvm()`.
+
+[source, C]
+----
+enum tvm_memory_region_type {
+    /*
+     * Reserved for mapping confidential pages. The region is initially unpopulated, and pages
+     * of confidential memory can be inserted by calling `sbi_tee_host_add_tvm_zero_pages()` and
+     * `sbi_tee_host_add_tvm_measured_pages().
+     */
+    CONFIDENTIAL_REGION = 0,
+    /*
+     * The region is initially unpopulated, and pages of shared memory may be inserted by calling
+     * `sbi_tee_host_add_tvm_shared_pages()`. Attempts by a TVM VCPU to access an unpopulated region
+     * will cause a `SHARED_PAGE_FAULT` exit from `sbi_tee_host_run_tvm_vcpu()`.
+     */
+    SHARED_MEMORY_REGION = 1,
+    /*
+     * The region is unpopulated; attempts by a TVM VCPU to access this region will cause a
+     * `MMIO_PAGE_FAULT` exit from `sbi_tee_host_run_tvm_vcpu()`.
+     */
+    EMULATED_MMIO_REGION = 2,
+};
+----
+
+The possible error codes returned in `sbiret.error` are shown below.
+
+[#table_sbi_tee_host_add_tvm_memory_region_errors]
+.TEE Host Add TVM Memory Region
+[cols="2,3", width=90%, align="center", options="header"]
+|===
+| Error code              | Description
+| SBI_SUCCESS             | The operation completed successfully.
+| SBI_ERR_INVALID_ADDRESS | `tvm_gpa_addr` was invalid.
+| SBI_ERR_INVALID_PARAM   | `tvm_guest_id` or `region_len` were invalid, or the TVM wasn't
+                            in the correct state.
+| SBI_ERR_FAILED          | The operation failed for unknown reasons.
+|===
+
+=== Function: TEE Host Add TVM Page Table Pages (FID #9)
+[source, C]
+-----
+struct sbiret sbi_tee_host_add_tvm_page_table_pages(unsigned long tvm_guest_id,
+                                                    unsigned long base_page_address,
+                                                    unsigned long num_pages);
+-----
+Adds `num_pages` confidential memory starting at `base_page_address` to the
+TVM's page-table page-pool. The implied page size is 4KiB.
+
+Page table pages may be added at any time, and a typical usecase is in response to a TVM page fault.
+
+The possible error codes returned in `sbiret.error` are shown below.
+
+[#table_sbi_tee_host_add_tvm_page_table_pages_errors]
+.TEE Host Add TVM Page Table Pages
+[cols="2,3", width=90%, align="center", options="header"]
+|===
+| Error code              | Description
+| SBI_SUCCESS             | The operation completed successfully.
+| SBI_ERR_INVALID_ADDRESS | `base_page_address` was invalid.
+| SBI_ERR_INVALID_PARAM   | `tvm_guest_id` or `num_pages` were invalid,
+                             or `tsm_page_type` is invalid.
```

Comment:
| SBI_ERR_OUT_OF_PTPAGES | The operation could not complete due to insufficient page table pages.